### PR TITLE
[SPARK-28793][DOC][SQL] Document CREATE FUNCTION in SQL Reference

### PR DESCRIPTION
**Pull request by dilipbiswal**
### What changes were proposed in this pull request?
Document CREATE FUNCTION statement in SQL Reference Guide. 

### Why are the changes needed?
Currently Spark lacks documentation on the supported SQL constructs causing
confusion among users who sometimes have to look at the code to understand the
usage. This is aimed at addressing this issue.

### Does this PR introduce any user-facing change?
Yes. 

**Before:**
There was no documentation for this.

**After.**
Screen Shot 2019-09-22 at 3 01 52 PM
Screen Shot 2019-09-22 at 3 02 11 PM
Screen Shot 2019-09-22 at 3 02 39 PM
Screen Shot 2019-09-22 at 3 04 04 PM

### How was this patch tested?
Tested using jykyll build --serve

### CODE REVIEW

File: docs/sql-ref-syntax-ddl-create-function.md [hunk 1]
```diff
@@ -19,4 +19,148 @@ license: |
   limitations under the License.
 ---
 
-**This page is under construction**
+### Description
+`CREATE FUNCTION` statement is used to create a temporary or permanent function
+in Spark. Temporary functions are scoped at a session level where as permanent
+functions are created in the persistent catalog and are made available to
+all sessions. The resources specified in the `USING` clause are made available
+to all executors when they are executed for the first time.
```
Comment (gatorsmile) [inline docs/sql-ref-syntax-ddl-create-function.md]: We also need to explain we can create/register temporary SQL functions via Python/Scala/Java APIs.

File: docs/sql-ref-syntax-ddl-create-function.md [hunk 2]
```diff
@@ -19,4 +19,148 @@ license: |
...
+  <dd>
+    If specified, the resources for function are reloaded. This is mainly useful
+    to pick up any changes made to the implementation of the function. This
+    parameter is mutually exclusive to <code>IF NOT EXISTS</code> and can not
+    be specified together.
+  </dd>
+  <dt><code><em>TEMPORARY</em></code></dt>
+  <dd>
+    Indicates the scope of function being created. When TEMPORARY is specified, the
+    created function is valid in the current session. No persistent entry is made
```
[the remark this answers is not in the document]
Comment (dilipbiswal) [reply docs/sql-ref-syntax-ddl-create-function.md]: @gatorsmile thanks.. fixed.

File: docs/sql-ref-syntax-ddl-create-function.md [hunk 3]
```diff
@@ -19,4 +19,148 @@ license: |
...
+    <b>Syntax:</b>
+      <code>
+        USING { { (JAR | FILE ) resource_uri} , ...}
+      </code>
+  </dd>
+</dl>
+
+### Examples
+{% highlight sql %}
+-- 1. Create a simple UDF `SimpleUdf` that adds the supplied integet value by 10.
```
[the remark this answers is not in the document]
Comment (dilipbiswal) [reply docs/sql-ref-syntax-ddl-create-function.md]: @gatorsmile fixed.

File: docs/sql-ref-syntax-ddl-create-function.md [hunk 4]
```diff
@@ -19,4 +19,148 @@ license: |
...
+    optionally qualified with a database name. <br><br>
+    <b>Syntax:</b>
+      <code>
+        [database_name.]function_name
+      </code>
+  </dd>
+  <dt><code><em>class_name</em></code></dt>
+  <dd>
+    Specifies the name of the class that provides the implementation for function to be created.
+    The implementing class should extend from one of the base classes in `Hive` as follows:
```
[the remark this answers is not in the document]
Comment (dilipbiswal) [reply docs/sql-ref-syntax-ddl-create-function.md]: @gatorsmile I have created a place holder link for custom scalar functions. There is already a place holder for aggregate functions and a lot of content is already in place.

File: docs/sql-ref-syntax-ddl-create-function.md (hunk 1 above)
Comment (dilipbiswal) [reply docs/sql-ref-syntax-ddl-create-function.md]: @gatorsmile I have created a place holder link for custom scalar functions. There is already a place holder for aggregate functions and a lot of content is already in place.

File: docs/sql-ref-syntax-ddl-create-function.md [hunk 5]
```diff
@@ -19,4 +19,154 @@ license: |
...
+|                   21|
+|                   22|
++---------------------+
+
+{% endhighlight %}
+
+### Related statements
+- [SHOW FUNCTIONS](sql-ref-syntax-aux-show-functions.html)
+- [DESCRIBE FUNCTION](sql-ref-syntax-aux-describe-function.html)
+- [DESCRIBE FUNCTION](sql-ref-syntax-aux-describe-function.html)
```
Comment (gatorsmile) [inline docs/sql-ref-syntax-ddl-create-function.md]: duplicate?

File: docs/sql-getting-started.md [hunk 6]
```diff
@@ -346,6 +346,9 @@ For example:
 
 </div>
 
+## Scalar Functions
+(to be filled soon)
```
Comment (gatorsmile) [inline docs/sql-getting-started.md]: Create a JIRA?

**Review by gatorsmile — APPROVED** [review]
LGTM except two comments.

File: docs/sql-ref-syntax-ddl-create-function.md (hunk 5 above)
Comment (dilipbiswal) [reply docs/sql-ref-syntax-ddl-create-function.md]: @gatorsmile oops.. thanks for catching it. Will fix.

File: docs/sql-getting-started.md (hunk 6 above)
Comment (dilipbiswal) [reply docs/sql-getting-started.md]: @gatorsmile created [here](https://issues.apache.org/jira/browse/SPARK-29458)

**Comment by dilipbiswal** [comment]
cc @gatorsmile

File: docs/sql-ref-syntax-ddl-create-function.md [hunk 7]
```diff
@@ -19,4 +19,153 @@ license: |
...
+  <dt><code><em>TEMPORARY</em></code></dt>
+  <dd>
+    Indicates the scope of function being created. When TEMPORARY is specified, the
+    created function is valid and visible in the current session. No persistent
+    entry is made in the catalog for these kind of functions.
+  </dd>
+  <dt><code><em>IF NOT EXISTS</em></code></dt>
+  <dd>
+    If specified, creates the function only when it does not exist. The creation
+    of function succeeds (no error is thrown), if the specified function already
```
Comment (srowen) [inline docs/sql-ref-syntax-ddl-create-function.md]: Nit: remove comma

File: docs/sql-ref-syntax-ddl-create-function.md [hunk 8]
```diff
@@ -19,4 +19,153 @@ license: |
...
+  <dt><code><em>OR REPLACE</em></code></dt>
+  <dd>
+    If specified, the resources for function are reloaded. This is mainly useful
+    to pick up any changes made to the implementation of the function. This
+    parameter is mutually exclusive to <code>IF NOT EXISTS</code> and can not
+    be specified together.
+  </dd>
+  <dt><code><em>TEMPORARY</em></code></dt>
+  <dd>
+    Indicates the scope of function being created. When TEMPORARY is specified, the
```
Comment (srowen) [inline docs/sql-ref-syntax-ddl-create-function.md]: Nit: code-format `TEMPORARY`?

File: docs/sql-ref-syntax-ddl-create-function.md [hunk 9]
```diff
@@ -19,4 +19,153 @@ license: |
...
+{% highlight sql %}
+CREATE [ OR REPLACE ] [ TEMPORARY ] FUNCTION [ IF NOT EXISTS ]
+    function_name AS class_name [ resource_locations ]
+{% endhighlight %}
+
+### Parameters
+<dl>
+  <dt><code><em>OR REPLACE</em></code></dt>
+  <dd>
+    If specified, the resources for function are reloaded. This is mainly useful
```
Comment (srowen) [inline docs/sql-ref-syntax-ddl-create-function.md]: for the function

File: docs/sql-ref-syntax-ddl-create-function.md [hunk 10]
```diff
@@ -19,4 +19,153 @@ license: |
...
+    optionally qualified with a database name. <br><br>
+    <b>Syntax:</b>
+      <code>
+        [database_name.]function_name
+      </code>
+  </dd>
+  <dt><code><em>class_name</em></code></dt>
+  <dd>
+    Specifies the name of the class that provides the implementation for function to be created.
+    The implementing class should extend from one of the base classes as follows:
```
Comment (srowen) [inline docs/sql-ref-syntax-ddl-create-function.md]: remove 'from'

File: docs/sql-ref-syntax-ddl-create-function.md [hunk 11]
```diff
@@ -19,4 +19,153 @@ license: |
...
+    <b>Syntax:</b>
+      <code>
+        USING { { (JAR | FILE ) resource_uri} , ...}
+      </code>
+  </dd>
+</dl>
+
+### Examples
+{% highlight sql %}
+-- 1. Create a simple UDF `SimpleUdf` that adds the supplied integral value by 10.
```
Comment (srowen) [inline docs/sql-ref-syntax-ddl-create-function.md]: "increments" rather than "adds"?

File: docs/sql-ref-syntax-ddl-create-function.md [hunk 12]
```diff
@@ -19,4 +19,153 @@ license: |
...
+--    import org.apache.hadoop.hive.ql.exec.UDF;
+--    public class SimpleUdf extends UDF {
+--      public int evaluate(int value) {
+--      return value + 10;
```
Comment (srowen) [inline docs/sql-ref-syntax-ddl-create-function.md]: Nit: indent this more

File: docs/sql-ref-syntax-ddl-create-function.md [hunk 13]
```diff
@@ -19,4 +19,153 @@ license: |
...
+--      }
+--    }
+-- 2. Compile and place it in a jar file called `SimpleUdf.jar` in /tmp.
```
Comment (srowen) [inline docs/sql-ref-syntax-ddl-create-function.md]: Nit: jar -> JAR

File: docs/sql-ref-syntax-ddl-create-function.md [hunk 14]
```diff
@@ -19,4 +19,153 @@ license: |
...
+-- database associated with it.
+SHOW USER FUNCTIONS;
+  +------------------+
+  |          function|
+  +------------------+
+  |default.simple_udf|
+  |   simple_temp_udf|
+  +------------------+
+
+-- 1. Mofify `SimpleUdf`'s implementation to add supplied integral value by 20.
```
Comment (srowen) [inline docs/sql-ref-syntax-ddl-create-function.md]: Mofify -> Modify

File: docs/sql-ref-syntax-ddl-create-function.md [hunk 15]
```diff
@@ -19,4 +19,153 @@ license: |
...
+--    import org.apache.hadoop.hive.ql.exec.UDF;
+  
+--    public class SimpleUdfR extends UDF {
+--      public int evaluate(int value) {
+--      return value + 20;
```
Comment (srowen) [inline docs/sql-ref-syntax-ddl-create-function.md]: Indent here

**Review by srowen — CHANGES_REQUESTED** [review]

**Review by srowen — APPROVED** [review]

**Comment by srowen** [comment]
Merged to master

**Comment by dilipbiswal** [comment]
Thanks a lot @srowen @gatorsmile

**Pull request closed without merging**